### PR TITLE
IMB-104: Add Cron Job for db table replacer

### PR DESCRIPTION
**Pull request by adityababumallisettiHO**
## What?
This work is part of tickets
https://collaboration.homeoffice.gov.uk/jira/browse/IMB-104
[https://collaboration.homeoffice.gov.uk/jira/browse/IMB-68](url)
CronJob is meant for performing regular scheduled actions. It runs a Job periodically on a given schedule, written in [Cron](https://en.wikipedia.org/wiki/Cron) format.
This Cron job Runs Automated Tasks at midnight and is documented in the confluence [https://collaboration.homeoffice.gov.uk/display/DSASS/High+Level+Ellaboration+-+IMB-68](url)
Link for Solution Design for Retrieving, processing data and notifying stakeholders: [[IMA Midnight Process Diagram](https://homeofficegovuk.sharepoint.com/:b:/r/sites/hof-team/Shared%20Documents/IMA/diagrams/midnight-process.pdf?csf=1&web=1&e=UTrURE)]
## Why?
Refer to https://collaboration.homeoffice.gov.uk/display/DSASS/High+Level+Ellaboration+-+IMB-68
## How?
Cron jobs require a config file. 
Cron job has been deployed with this pipeline https://drone-gh.acp.homeoffice.gov.uk/UKHomeOffice/ima/867
## Testing?
For testing We have enabled cronjob in branch namespace temporarily. We will remove this cronjob from Branch before merging to Master Branch. 
We will only need this cron job in UAT and PROD

## Screenshots (optional)
image

## Anything Else? (optional)
The quick and simple editor for cron schedule expressionshttps://crontab.guru/
image

Commands to list the Cron Jobs
kubectl --context=acp-notprod_SAS --namespace=sas-ima-branch get cronjob

### CODE REVIEW

File: kube/cron/hof_db_table_replacer.yaml [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+            securityContext:
+              runAsNonRoot: true
+            envFrom:
+              - configMapRef:
+                  {{ if eq .KUBE_NAMESPACE .BRANCH_ENV }}
+                  name: configmap-{{ .DRONE_SOURCE_BRANCH }}
+                  {{ else }}
+                  name: configmap
+                  {{ end }}
+            env:
```
Comment (jamiecarterHO) [inline kube/cron/hof_db_table_replacer.yaml:36]: I will send you a list of env vars we will need here. There are a few more than initially in this file.

File: kube/cron/hof_db_table_replacer.yaml [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+                valueFrom:
+                  secretKeyRef:
+                    name: keycloak-user
+                    key: username
+              - name: KEYCLOAK_PASSWORD
+                valueFrom:
+                  secretKeyRef:
+                    name: keycloak-user
+                    key: password
+            command: ["yarn", "run", "generate:reports"]
```
Comment (jamiecarterHO) [inline kube/cron/hof_db_table_replacer.yaml:76]: Maybe we should remove/comment this command and see if the Dockerfile will run the script in its own steps. If so we won't need a command property here.

File: kube/cron/hof_db_table_replacer.yaml [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+              - name: NOTIFY_KEY
+                valueFrom:
+                  secretKeyRef:
+                    name: notify-key
+                    key: notify-key
+              - name: FILE_VAULT_URL
+              {{ if eq .KUBE_NAMESPACE .PROD_ENV }}
+                value: fv-ima.sas.homeoffice.gov.uk/file
+              {{ else if eq .KUBE_NAMESPACE .STG_ENV }}
+                value: fv-ima.stg.sas.homeoffice.gov.uk/file
```
Comment (jamiecarterHO) [inline kube/cron/hof_db_table_replacer.yaml:50]: I think we can get rid of the staging value for FILE_VAULT_URL

File: kube/cron/hof_db_table_replacer.yaml [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+              - name: DB_REPLACER_SOURCE_FILE_TABLE
+                valueFrom:
+                  configMapKeyRef:
+                    {{ if eq .KUBE_NAMESPACE .BRANCH_ENV }}
+                    name: configmap-{{ .DRONE_SOURCE_BRANCH }}
+                    {{ else }}
+                    name: configmap
+                    {{ end }}
+                    key: DB_REPLACER_SOURCE_FILE_TABLE
+
```
Comment (jamiecarterHO) [inline kube/cron/hof_db_table_replacer.yaml:94]: These two variables will already be in the environment from the config map (lines 29-35). The ones we need to add to this manifest are:

```bash
- name: DB_HOST
    valueFrom:
      secretKeyRef:
        key: endpoint
        {{ if or (eq .KUBE_NAMESPACE .BRANCH_ENV) (eq .KUBE_NAMESPACE .UAT_ENV) }}
        name: ima-notprod-rds
        {{ else }}
        name: ima-prod
        {{ end }}
  - name: DB_NAME
    valueFrom:
      secretKeyRef:
        key: db_name
        {{ if or (eq .KUBE_NAMESPACE .BRANCH_ENV) (eq .KUBE_NAMESPACE .UAT_ENV) }}
        name: ima-notprod-rds
        {{ else }}
        name: ima-prod
        {{ end }}
  - name: DB_USER
    valueFrom:
      secretKeyRef:
        key: username
        {{ if or (eq .KUBE_NAMESPACE .BRANCH_ENV) (eq .KUBE_NAMESPACE .UAT_ENV) }}
        name: ima-notprod-rds
        {{ else }}
        name: ima-prod
        {{ end }}
  - name: DB_PASS
    valueFrom:
      secretKeyRef:
        key: password
        {{ if or (eq .KUBE_NAMESPACE .BRANCH_ENV) (eq .KUBE_NAMESPACE .UAT_ENV) }}
        name: ima-notprod-rds
        {{ else }}
        name: ima-prod
        {{ end }}
  - name: DB_PORT
    value: "5432"
  - name: DB_INSERT_BATCH_SIZE
    value: "2000"
  - name: DB_CLIENT
    value: knex
  - name: DB_MODEL
    value: knex-postgres-model
  - name: SERVICE_NAME
    value: ima
```

File: kube/cron/hof_db_table_replacer.yaml (hunk 2 above)
Comment (adityababumallisettiHO) [reply kube/cron/hof_db_table_replacer.yaml]: Commented out this command .

File: kube/cron/hof_db_table_replacer.yaml (hunk 4 above)
Comment (adityababumallisettiHO) [reply kube/cron/hof_db_table_replacer.yaml]: Hi @jamiecarterHO , Thanks for providing me these information. I have updated it now.

**Review by jamiecarterHO — APPROVED** [review]
lgtm

**Pull request merged**